Sources/CardPayments/APIRequests/ConfirmPaymentSourceRequest.swift

```diff
@@ -4,132 +4,89 @@ import CorePayments
-        
-        var paymentSource: PaymentSource?
-        var applicationContext: ApplicationContext?
+    private enum PaymentSourceKeys: String, CodingKey {
+        case card
     }
     
-    private struct ApplicationContext: Encodable {
-        
-        let returnUrl: String
-        let cancelUrl: String
-    }
-    
-    enum CodingKeys: String, CodingKey {
+    private enum CardKeys: String, CodingKey {
         case number
         case expiry
-        case securityCode
-        case name
         case billingAddress
+        case name
         case attributes
     }
     
-    private struct EncodedCard: Encodable {
-        
-        let number: String
-        let securityCode: String
-        let billingAddress: Address?
-        let name: String?
-        let expiry: String
-        let attributes: Attributes?
-                
-        func encode(to encoder: Encoder) throws {
-            var container = encoder.container(keyedBy: CodingKeys.self)
-            try container.encode(number, forKey: .number)
-            try container.encode(expiry, forKey: .expiry)
-            try container.encode(securityCode, forKey: .securityCode)
-            try container.encode(name, forKey: .name)
-            try container.encode(billingAddress, forKey: .billingAddress)
-            try container.encode(attributes, forKey: .attributes)
-        }
+    private enum BillingAddressKeys: String, CodingKey {
+        case addressLine1 = "address_line_1"
+        case addressLine2 = "address_line_2"
+        case adminArea1 = "admin_area_1"
+        case adminArea2 = "admin_area_2"
+        case postalCode
+        case countryCode
     }
     
-    private struct Verification: Encodable {
-        
-        let method: String
+    private enum AttributesKeys: String, CodingKey {
+        case vault
+        case verification
     }
     
-    private struct Attributes: Encodable {
-        
-        let verification: Verification
-        
-        init(verificationMethod: String) {
-            self.verification = Verification(method: verificationMethod)
-        }
+    private enum VerificationKeys: String, CodingKey {
+        case method
+    }
+    
+    // MARK: - Private Properties
+    
+    private let returnURL = PayPalCoreConstants.callbackURLScheme + "://card/success"
+    private let cancelURL = PayPalCoreConstants.callbackURLScheme + "://card/cancel"
+    private let cardRequest: CardRequest
+    
+    // MARK: - Initializer
+    
+    init(cardRequest: CardRequest) {
+        self.cardRequest = cardRequest
     }
     
-    private struct PaymentSource: Encodable {
+    // MARK: - Custom Encoder
+    
+    func encode(to encoder: Encoder) throws {
+        var topLevel = encoder.container(keyedBy: TopLevelKeys.self)
         
-        let card: EncodedCard
+        var applicationContext = topLevel.nestedContainer(keyedBy: ApplicationContextKeys.self, forKey: .applicationContext)
```

Comment:
It's mainly just for readability. If we use the default, then we have to format our structs & properties to be nested in the same exact format that we want the request to be encoded as ([see previous implementation](https://github.com/paypal/paypal-ios/blob/main/Sources/CardPayments/APIRequests/ConfirmPaymentSourceRequest.swift)).

Whereas with the custom encoder style ([see file](https://github.com/paypal/paypal-ios/blob/construct-url-request-in-http/Sources/CardPayments/APIRequests/ConfirmPaymentSourceRequest.swift)), we can basically flatten all of our properties & keys and IMO this init is easier to digest the format of the POST versus the nested structs.

Reply:
Also, we were doing the custom encoder already to format the[ analytics post body ](https://github.com/paypal/paypal-ios/blob/main/Sources/CorePayments/AnalyticsEventData.swift), for the same reason to avoid needing to nest a bunch of structs and properties just to satisfy the API format requirement.

Let me know if that makes sense!

Reply:
Yeah it makes sense for the most part. This [implementation](https://github.com/paypal/paypal-ios/blob/main/Sources/CardPayments/APIRequests/ConfirmPaymentSourceRequest.swift) has some additional logic also besides `Codable` conformance. I'm wondering if taking out all of the default value initialization would allow us to keep the default codable implementation. That would be a win if we could write less code.

Reply:
Yep - all of that extra logic was removed from [`ConfirmPaymentSourceRequest`](https://github.com/paypal/paypal-ios/blob/construct-url-request-in-http/Sources/CardPayments/APIRequests/ConfirmPaymentSourceRequest.swift) (in a future PR we're gonna nuke that `APIRequest` protocol entirely 🔥). So in this PR we now have data objects that _only_ represent the model type which is nice. It doesn't do anything more.

If we want the flat structure though, and to avoid those nested structs, we will need to implement `encode()`.